spec/std/uri_spec.cr

```diff
@@ -1,51 +1,143 @@
 require "spec"
 require "uri"
 
-private def assert_uri(string, scheme = nil, host = nil, port = nil, path = "", query = nil, user = nil, password = nil, fragment = nil, opaque = nil)
-  it "parse #{string}" do
-    uri = URI.parse(string)
-    uri.scheme.should eq(scheme)
-    uri.host.should eq(host)
-    uri.port.should eq(port)
-    uri.path.should eq(path)
-    uri.query.should eq(query)
-    uri.user.should eq(user)
-    uri.password.should eq(password)
-    uri.fragment.should eq(fragment)
-    uri.opaque.should eq(opaque)
+private def assert_uri(string, file = __FILE__, line = __LINE__, **args)
+  it "`#{string}`", file, line do
+    URI.parse(string).should eq URI.new(**args)
+    URI.parse(string).to_s.should eq string
   end
 end
 
 describe "URI" do
-  assert_uri("http://www.example.com", scheme: "http", host: "www.example.com")
-  assert_uri("http://www.example.com:81", scheme: "http", host: "www.example.com", port: 81)
-  assert_uri("http://[::1]:81", scheme: "http", host: "[::1]", port: 81)
-  assert_uri("http://www.example.com/foo", scheme: "http", host: "www.example.com", path: "/foo")
-  assert_uri("http://www.example.com/foo?q=1", scheme: "http", host: "www.example.com", path: "/foo", query: "q=1")
-  assert_uri("http://www.example.com?q=1", scheme: "http", host: "www.example.com", query: "q=1")
-  assert_uri("https://www.example.com", scheme: "https", host: "www.example.com")
-  assert_uri("https://alice:pa55w0rd@www.example.com", scheme: "https", host: "www.example.com", user: "alice", password: "pa55w0rd")
-  assert_uri("https://alice@www.example.com", scheme: "https", host: "www.example.com", user: "alice", password: nil)
-  assert_uri("https://%3AD:%40_%40@www.example.com", scheme: "https", host: "www.example.com", user: ":D", password: "@_@")
-  assert_uri("https://www.example.com/#top", scheme: "https", host: "www.example.com", path: "/", fragment: "top")
-  assert_uri("http://www.foo-bar.example.com", scheme: "http", host: "www.foo-bar.example.com")
-  assert_uri("/foo", path: "/foo")
-  assert_uri("/foo?q=1", path: "/foo", query: "q=1")
-  assert_uri("mailto:foo@example.org", scheme: "mailto", path: nil, opaque: "foo@example.org")
+  describe ".parse" do
+    # scheme
+    assert_uri("http:", scheme: "http")
+    it { URI.parse("HttP:").should eq(URI.new(scheme: "http")) }
```

Comment:
you are missing a spec of what the result of `#scheme` on a missing scheme is. Some tests below lacks scheme but none tests what `#scheme` is in that case.

Reply:
I'm not sure what you mean. There are many specs for no-scheme URIs. For example `assert_uri("/foo", path: "/foo")`. It ensures that the URI parsed from `"/foo"` has a path of `/foo` and all other properties at the default value. This means `scheme` should be `nil`.